src/v/kafka/server/fetch_session.h

```diff
@@ -100,7 +100,9 @@ class fetch_partitions_linked_hash_map {
 
     static auto make_partition_iterator(io_list_t::const_iterator it) {
         return boost::iterators::make_transform_iterator(
-          it, [](const entry& e) { return e.partition; });
+          it, [](const entry& e) -> const kafka::fetch_session_partition& {
```

Comment:
This small change was a big win, like 20% or something: without it we return the e.partition (a fairly big structure, including dynamic allocations) by _value_ resulting in a copy. Iterators generally return a reference (as we do now) for a good reason!

Reply:
>This small change was a big win, like 20% or something:

oh this is really great. how did you find it? seeing the allocations / hotspot with perf?

Reply:
Exactly I was using `perf record` and `report report` and looking at the mixed source-assembly view and I could see a lot of time in the string copy loop in the middle of iteration which made me take a look at the iterator. Because `PERF_TEST` offers a count of allocations during the test, it was easy to quickly verify that this fix dropped the allocation count by 1 (which was substantial since there were only 4 allocation total in the inner loop).

This is kind of a brute force method though and with the memory profiler being added by @StephanDollberg there might be easier ways too, e.g., "enable profiling for this benchmark and get a dump of all allocation sites".